go/src/koding/kites/kloud/machine/mongo.go

```diff
@@ -0,0 +1,179 @@
+					machinesDB = append(machinesDB[:i], machinesDB[i+1:]...)
+					i--
+				}
+			}
+		}
+	}
+
+	// Get stack template IDs used by machines. Using map as the storage will
+	// remove duplicated values.
+	stackTmplIDs := make(map[bson.ObjectId]struct{})
+	for i := range machinesDB {
+		stackTmplIDs[machinesDB[i].GeneratedFrom.TemplateId] = struct{}{}
+	}
+
+	// We don't need to search in jGroups collection in order to find team name
+	// because jStackTemplates also contains that information.
+	stackTmplsDB, err := m.adapter.GetStackTemplateFieldsByIds(
+		toSlice(stackTmplIDs),             // stack templates to find.
+		[]string{"_id", "group", "title"}, // fields we need from jStackTemplates.
+	)
+	if err != nil {
+		return nil, models.ResError(err, modelhelper.StackTemplateColl)
+	}
+
+	// Helper made to simplify searching for group and title names.
+	groupTitles := make(map[bson.ObjectId][2]string, len(stackTmplsDB))
+	for _, st := range stackTmplsDB {
+		groupTitles[st.Id] = [2]string{st.Group, st.Title}
+	}
+
+	machines := make([]*Machine, len(machinesDB))
+	for i, mdb := range machinesDB {
+		machines[i] = &Machine{
+			Team:      groupTitles[mdb.GeneratedFrom.TemplateId][0],
+			Stack:     groupTitles[mdb.GeneratedFrom.TemplateId][1],
+			Provider:  mdb.Provider,
+			Label:     mdb.Label,
+			IP:        hostOnly(mdb.IpAddress),
+			CreatedAt: mdb.CreatedAt,
+			Status: Status{
+				State:      mdb.Status.State,
+				Reason:     mdb.Status.Reason,
+				ModifiedAt: mdb.Status.ModifiedAt,
+			},
+			Users: filterUsers(mdb.Users, f),
+		}
+	}
+
+	return machines, nil
+}
+
+// toSlice is a helper function that converts bson.ObjectId set to slice.
+func toSlice(set map[bson.ObjectId]struct{}) (s []bson.ObjectId) {
+	for objID, _ := range set {
+		s = append(s, objID)
+	}
+	return s
+}
+
+// makeMachineUser converts machine users from model to MachineUser object.
+// We have separate MachineUser type here to not propagate internal
+// bson.ObjectId value outside the back-end.
+func makeMachineUser(mu *models.MachineUser) User {
+	return User{
+		Sudo:      mu.Sudo,
+		Owner:     mu.Owner,
+		Permanent: mu.Permanent,
+		Approved:  mu.Approved,
+		Username:  mu.Username,
+	}
+}
+
+// filterUsers removes users specified by provided filter. This prevents from
+// receiving possibly sensitive information about machine users.
+func filterUsers(users []models.MachineUser, f *Filter) (res []User) {
+	for i := range users {
+		// Filter by user name.
+		if f.Username != "" && users[i].Username == f.Username {
+			res = append(res, makeMachineUser(&users[i]))
+			continue
+		}
+
+		// Keep machine owners in order to know who shared his machine with us.
+		if f.Owners && users[i].Owner && users[i].Sudo {
+			res = append(res, makeMachineUser(&users[i]))
+			continue
+		}
+	}
+
+	return
+}
+
+// hostOnly returns only host part of provided address. This is a fix for
+// machine IpAddress field which contains klient's port.
```

Comment:
Sorry, I forgot that existing documents are going to still have host:port. While the info command will slowly fix all documents, until that happens the workaround still is going to be needed.